arborist/user.go

```diff
@@ -150,6 +150,56 @@ func userWithName(db *sqlx.DB, name string) (*UserFromQuery, error) {
 	return &user, nil
 }
 
+type PolicyInfoFromQuery struct {
+	Name           string `db:"name"`
+	Authz_provider string `db:"authz_provider"`
+	Type           string `db:"type"`
+}
+
+func fetchPolicyInfo(db *sqlx.DB, user_name string, policy_name string) (*PolicyInfoFromQuery, error) {
+
+	stmt := `
+	SELECT * from
+	(
+		SELECT policy.name AS name, usr_policy.authz_provider as authz_provider, 'direct' as type
+			FROM usr_policy
+			INNER JOIN policy ON policy.id = usr_policy.policy_id
+			INNER JOIN usr on usr_policy.usr_id = usr.id 
+			WHERE usr.name like $1
+		UNION
+		SELECT policy.name AS name, grp_policy.authz_provider as authz_provider, 'group' as type
+			FROM usr_grp
+			INNER JOIN grp_policy ON grp_policy.grp_id = usr_grp.grp_id
+			INNER JOIN policy ON policy.id = grp_policy.policy_id
+			INNER JOIN usr on usr_grp.usr_id = usr.id
+			WHERE usr.name like $1
+		UNION
+		SELECT policy.name AS name, grp_policy.authz_provider as authz_provider, 'group' as type
+			FROM grp
+			INNER JOIN grp_policy ON grp_policy.grp_id = grp.id
+			INNER JOIN policy ON policy.id = grp_policy.policy_id
+			WHERE grp.name IN ( $3 , $4)
+	) as all_policies where name like $2;
+	`
+	policyInfoList := []PolicyInfoFromQuery{}
+	err := db.Select(
+		&policyInfoList,
+		stmt,
+		user_name,      // $1
+		policy_name,    // $2
+		AnonymousGroup, // $3
+		LoggedInGroup,  // $4
+	)
+	if err != nil {
+		return nil, err
+	}
+	if len(policyInfoList) == 0 {
+		return nil, nil
+	}
+	policyInfo := policyInfoList[0]
```

Comment:
i think there's a problem with this logic if the same policy was granted to the user via 1 or more groups, as well as directly:
- if the first returned row is "group", you'll refuse to revoke even though you could revoke the direct policy access.
- group access and direct access might have been granted via different authz providers, which makes the returned data confusing

I think we can fix this and at the same time simplify this function, by _not_ querying the groups. _And_ this function will be reusable when we implement PXP-9148! From the design doc for PXP-9148:

> Expose authz_provider data: add a GET /user/username/policy/policy_id endpoint that queries the usr_policy table and returns:
```
{
  usr_id:
  policy_id:
  expires_at:
  authz_provider:
}
```
> Return 404 if we fail to query the usr_policy table, which means the access was granted via a group.

You could update `fetchPolicyInfo` to do the above (except no need to return the data you don't need for this ticket). So if `fetchPolicyInfo` does not return policy info, your calling function can return `400: either user does not have policy X, or it was granted via a group`.

And there should only be 1 row with 1 authz provider, since we're only fetching direct access. If there are more than 1 row, we might be looking at an edge case, and you should probably log a warning